### PR TITLE
Feature/csl278 blocks plain files

### PR DESCRIPTION
**Pull request by pva701**
Blocks are stored in plain files now - file per block. We should store several blocks in one file in the future.

### CODE REVIEW

File: core/Pos/Util/Mnemonics.hs [hunk 1]
```diff
@@ -125,7 +124,7 @@ indicesToBS is = do
     when lrg $ Left "indicesToBS: index larger or equal than 2048"
     return . pad . integerToBS $ (foldl' f 0 is) `shiftL` shift_width
   where
-    lrg = not . isNothing $ find (>= 2048) is
```
Comment (volhovM) [inline core/Pos/Util/Mnemonics.hs]: This is funny.

File: core/Pos/Util/Mnemonics.hs [hunk 2]
```diff
@@ -57,7 +56,7 @@ toMnemonic ent = do
         Left "toMnemonic: maximum entropy is 64 bytes (512 bits)"
     when (isJust $ find (not . isAscii) ms) $
         Left "fromMnemonic: non-ASCII characters not supported"
-    return $ T.pack ms
+    return $ toText ms
```
Comment (volhovM) [inline core/Pos/Util/Mnemonics.hs]: Why not use `fromString`? Is there any rule of thumb for these situations?

File: src/node/Main.hs [hunk 3]
```diff
@@ -135,7 +136,9 @@ userSecretWithGenesisKey Args{..} userSecret
           Nothing -> fetchPrimaryKey userSecret
           Just i -> do
               let sk = genesisDevSecretKeys !! i
-                  us = userSecret & usPrimKey .~ Just sk
+                  us = userSecret
+                       & usPrimKey .~ Just sk
+                       & usKeys %~ (noPassEncrypt sk :)
```
Comment (volhovM) [inline src/node/Main.hs]: How is this relevant?

Comment (gromakovsky) [inline src/node/Main.hs]: Someone doesn't know how to use git.

File: core/Pos/Util/Mnemonics.hs (hunk 2 above)
Comment (gromakovsky) [inline core/Pos/Util/Mnemonics.hs]: What is `fromString`? Is it from `IsString`? I guess it's not in `Universum`.

Comment (pva701) [inline core/Pos/Util/Mnemonics.hs]: it's not my code at all, but yes `fromString` ins'n in Universum

File: db/Pos/DB/Functions.hs [hunk 4]
```diff
@@ -40,7 +40,7 @@ import           Pos.DB.Error          (DBError (DBMalformed))
 import           Pos.DB.Iterator.Class (DBIteratorClass (..))
 import           Pos.DB.Types          (DB (..))
 
-openDB :: MonadIO m => FilePath -> m (DB)
+openDB :: MonadIO m => FilePath -> m DB
```
Comment (gromakovsky) [inline db/Pos/DB/Functions.hs]: Summer is coming 🌤  😎

Comment (pva701) [inline db/Pos/DB/Functions.hs]: )))))))))))))))))))))))))))))))))))

File: db/Pos/DB/Class.hs [hunk 5]
```diff
@@ -29,7 +29,7 @@ import           Pos.Util.Iterator            (ListHolderT (..))
 
 -- TODO write a documentation. LensLike' is just a lens. Written using
 -- LensLike' to avoid rankntypes.
-class (MonadIO m, MonadThrow m) => MonadDB m where
+class (MonadIO m, MonadThrow m, MonadCatch m) => MonadDB m where
```
Comment (gromakovsky) [inline db/Pos/DB/Class.hs]: IIRC, `MonadCatch` implies `MonadThrow`. Also I am not sure that it's good to have these constraints here.

File: db/Pos/DB/Types.hs [hunk 6]
```diff
@@ -36,11 +37,12 @@ data DB = DB
     }
 
 data NodeDBs = NodeDBs
-    { _blockDB  :: !DB     -- ^ Blocks, block index, undo data.
-    , _gStateDB :: !DB     -- ^ Global state corresponding to some tip.
-    , _lrcDB    :: !DB     -- ^ Data computed by LRC.
-    , _miscDB   :: !DB     -- ^ Everything small and insignificant
-    , _miscLock :: !RWLock -- ^ Lock on misc db
+    { _blockIndexDB  :: !DB       -- ^ Block index.
+    , _blockData     :: !FilePath -- ^ Block and undo files.
```
Comment (gromakovsky) [inline db/Pos/DB/Types.hs]: I find it slightly counter-intuitive that field which stores `FilePath` is called `…Data`. Maybe it's only me who thinks so.

File: src/Pos/DB/Block.hs [hunk 7]
```diff
@@ -207,15 +211,32 @@ undoKey h = "u" <> convert h
...
+getData = undefined
+
+putData ::  (MonadIO m, Bi v) => FilePath -> v -> m ()
+putData = undefined
+
+deleteData :: (MonadIO m) => FilePath -> m ()
+deleteData = undefined
+
+blockDataPath :: MonadDB m => HeaderHash -> m FilePath
+blockDataPath (unpack . sformat (hashHexF%".block") -> fn) =
```
Comment (gromakovsky) [inline src/Pos/DB/Block.hs]: 1. There is `formatToString`.
2. There is `toText`.

File: src/Pos/DB/Block.hs [hunk 8]
```diff
@@ -207,15 +214,39 @@ undoKey h = "u" <> convert h
...
+        Right r -> Just r
+    handle (_::SomeException) = pure Nothing
+
+putData ::  (MonadIO m, Bi v) => FilePath -> v -> m ()
+putData fp = liftIO . BS.writeFile fp . encodeStrict
+
+deleteData :: (MonadIO m, MonadCatch m) => FilePath -> m ()
+deleteData fp = (liftIO $ removeFile fp) `catch` handle
+  where
+    handle (_::SomeException) = pure () -- poh
```
Comment (gromakovsky) [inline src/Pos/DB/Block.hs]: I don't like such comments.

File: src/Pos/DB/Block.hs [hunk 9]
```diff
@@ -221,14 +224,21 @@ putBi k v = rocksPutBi k v =<< getBlockIndexDB
...
 
-getData ::  (MonadIO m, Bi v) => FilePath -> m (Maybe v)
-getData = undefined
+getData ::  (MonadIO m, MonadCatch m, Bi v) => FilePath -> m (Maybe v)
+getData fp = liftIO (decodeMaybe . BSL.fromStrict <$> BS.readFile fp) `catch` handle
+  where
+    decodeMaybe x = case decodeFull x of
+        Left _  -> Nothing
+        Right r -> Just r
+    handle (_::SomeException) = pure Nothing
```
Comment (gromakovsky) [inline src/Pos/DB/Block.hs]: I don't like it to be honest. We can't distinguish situations when there is no data/when it's malformed (can't be parsed)/when hard drive is broken. I think it's bad approach.

Comment (pva701) [inline src/Pos/DB/Block.hs]: I can add `throwM`

Comment (gromakovsky) [inline src/Pos/DB/Block.hs]: Why do you catch all exceptions here? I think caller should decide what to do if something goes wrong here.

File: db/Pos/DB/Types.hs (hunk 6 above)
Comment (volhovM) [inline db/Pos/DB/Types.hs]: Yes, I agree. Can it be `blocksDir` or something?

File: src/Pos/DB/Block.hs (hunk 9 above)
Comment (pva701) [inline src/Pos/DB/Block.hs]: do you agree to catch all exceptions in `deleteData` and don't catch them here7

File: core/Pos/Util/Mnemonics.hs (hunk 2 above)
Comment (ChShersh) [inline core/Pos/Util/Mnemonics.hs]: `toText` is also shorter

File: src/Pos/DB/Block.hs [hunk 10]
```diff
@@ -207,15 +211,32 @@ undoKey h = "u" <> convert h
...
+deleteData :: (MonadIO m) => FilePath -> m ()
+deleteData = undefined
+
+blockDataPath :: MonadDB m => HeaderHash -> m FilePath
+blockDataPath (unpack . sformat (hashHexF%".block") -> fn) =
+    (</> fn) . (^. blockData) <$> getNodeDBs
+
+undoDataPath :: MonadDB m => HeaderHash -> m FilePath
+undoDataPath (unpack . sformat (hashHexF%".undo") -> fn) =
+    (</> fn) . (^. blockData) <$> getNodeDBs
```
Comment (ChShersh) [inline src/Pos/DB/Block.hs]: @neongreen Can this be written more nicely?

File: src/Pos/DB/Block.hs [hunk 11]
```diff
@@ -221,14 +224,21 @@ putBi k v = rocksPutBi k v =<< getBlockIndexDB
 delete :: (MonadDB m) => ByteString -> m ()
 delete k = rocksDelete k =<< getBlockIndexDB
 
-getData ::  (MonadIO m, Bi v) => FilePath -> m (Maybe v)
-getData = undefined
+getData ::  (MonadIO m, MonadCatch m, Bi v) => FilePath -> m (Maybe v)
+getData fp = liftIO (decodeMaybe . BSL.fromStrict <$> BS.readFile fp) `catch` handle
+  where
+    decodeMaybe x = case decodeFull x of
```
Comment (ChShersh) [inline src/Pos/DB/Block.hs]: There is [`rightToMaybe`](https://github.com/serokell/universum/blob/f2ccf8afd862e37ccd204c0ef9efde48a05c2d29/src/Monad/Either.hs#L29) in `universum`

File: src/Pos/DB/Block.hs (hunk 10 above)
Comment (neongreen) [inline src/Pos/DB/Block.hs]: `getNodeDBs <&> \dbs -> dbs ^. blockData </> fn`

Comment (neongreen) [inline src/Pos/DB/Block.hs]: also `unpack . sformat` = `formatToString`

Comment (neongreen) [inline src/Pos/DB/Block.hs]: Also I find this variant to be most readable:

```haskell
undoDataPath hash = do
    blocksDir <- view blockData <$> getNodeDBs
    pure $ blocksDir </> formatToString (hashHexF%".undo") h
```

File: src/Pos/DB/Block.hs (hunk 9 above)
Comment (gromakovsky) [inline src/Pos/DB/Block.hs]: No, I don't understand what's the purpose of catching **all** exceptions in `deleteData`.

Comment (pva701) [inline src/Pos/DB/Block.hs]: @gromakovsky if file doesn't exist and we try to remove it, exception will be thrown

Comment (gromakovsky) [inline src/Pos/DB/Block.hs]: http://softwareengineering.stackexchange.com/a/189225

tl;dr: check whether file exists and remove it only if it does.

Comment (pva701) [inline src/Pos/DB/Block.hs]: otherwise throw exception7

Comment (gromakovsky) [inline src/Pos/DB/Block.hs]: Otherwise do nothing.

File: src/Pos/DB/Block.hs [hunk 12]
```diff
@@ -225,28 +225,25 @@ delete :: (MonadDB m) => ByteString -> m ()
 delete k = rocksDelete k =<< getBlockIndexDB
 
 getData ::  (MonadIO m, MonadCatch m, Bi v) => FilePath -> m (Maybe v)
-getData fp = liftIO (decodeMaybe . BSL.fromStrict <$> BS.readFile fp) `catch` handle
-  where
-    decodeMaybe x = case decodeFull x of
-        Left _  -> Nothing
-        Right r -> Just r
-    handle (_::SomeException) = pure Nothing
+getData fp = liftIO (rightToMaybe . decodeFull . BSL.fromStrict <$> BS.readFile fp)
```
Comment (gromakovsky) [inline src/Pos/DB/Block.hs]: No, I want to see an exception if decoding fails. That's what we do when we are working with rocksdb. Also I suspect that exception will be throw if file doesn't exist. But in this case I find it more reasonable to return `Nothing`.

Comment (pva701) [inline src/Pos/DB/Block.hs]: @gromakovsky fixed

File: src/Pos/DB/Block.hs [hunk 13]
```diff
@@ -225,7 +225,15 @@ delete :: (MonadDB m) => ByteString -> m ()
 delete k = rocksDelete k =<< getBlockIndexDB
 
 getData ::  (MonadIO m, MonadCatch m, Bi v) => FilePath -> m (Maybe v)
-getData fp = liftIO (rightToMaybe . decodeFull . BSL.fromStrict <$> BS.readFile fp)
+getData fp = flip catch handle $ liftIO $
+    either (\er -> throwM $ DBMalformed $ "Block deserialization failed: " <> show er) pure .
```
Comment (gromakovsky) [inline src/Pos/DB/Block.hs]: Why block? AFAIU, we also store `Undo`s in files and this `getData` can be applied to such file.

Comment (pva701) [inline src/Pos/DB/Block.hs]: @gromakovsky  done

**Comment by gromakovsky** [comment]
Approved, will merge after checks pass.
